src/tensors/cpu/wasm_intgemm_interface.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+/* Main interface for integer matrix multiplication (C = A * B) for wasm.
+ *
+ * A is typically activations whose rows should be a multiple of 1 (i.e. no restriction) and
+ * columns should be a multiple of 64.
+ * 
+ * B is typically fixed model parameters whose rows should be a multiple of 64 and columns
+ * should be a multiple of 8.
+ * 
+ * C is row major.
+ */
+
+#include <cstdint>
+
+using Index = unsigned int;
+
+
+/* Prepare B in a CPU-dependent format from an already quantized, transposed (routine not provided)
+ * and CPU-independent format of B. The prepared B will be used as an input to Multiply routine.
+ * This function is useful while using the quantized models that are stored in a CPU-independent
+ * format on the disk.
+ */
+void Int8PrepareBQuantizedTransposed(const int8_t* B_input,
+                                    Index B_untransposed_rows,
+                                    Index B_untransposed_cols,
+                                    int8_t* output);
+
+
+/* Select columns from a prepared B matrix. The number of selected columns must be a multiple of 8.
+ */
+void Int8SelectColumnsB(const int8_t* B_input,
+                        Index B_rows,
+                        Index B_cols,             // We need this for bound checking for B buffer?
+                        const Index* cols_begin,  // Shouldn't cols_begin be just Index and not Index*?
+                        const Index* cols_end,    // Shouldn't cols_end be just Index and not Index*?
```

Comment:
@XapaJIaMnu @kpu Could you please help me in understanding what SelectColumnsB function does?

Does it select columns of B from `cols_begin` to `cols_end` index with the condition that `0 <= cols_begin, cols_end < B_cols`

if yes, then shouldn't `cols_begin` and `cols_end` arguments be just plain `Index` instead of `Index*`?

Reply:
SelectColumsnB selects subset of columns from a matrix. So what you need is an array of indecis to select (let's say [1, 25, 8723, 9837]) this is why you need ptrs. In practise this is usually the `begin()` and `end()` of a vector: https://github.com/marian-nmt/marian-dev/blob/78e856e2d8a205587114071c8cfd9c2a0c7c7ebe/src/tensors/cpu/intgemm_interface.h#L213

Reply:
Okay. So the array of indices can also be represented as `(const Index* cols_begin, Index num_cols)`. In this case, `cols_end` will be equal to `cols_begin + num_cols - 1`

Thanks for the explanation.

Reply:
Let's keep it the same way it's implemented in the function to avoid unnecessary complexity.

Reply:
>  the array of indices can also be represented as (const Index* cols_begin, Index num_cols). In this case, cols_end will be equal to cols_begin + num_cols - 1

Actually it is just cols_end = cols_begin + num_cols.

There is no difference for underlined implementation (since second pointer is easy to calculate). In any case, you will have to know exact size of `Index` (not `unsigned int` of course) for layout of `cols_XXX` array.

Reply:
@yurydelendik is correct we are half-open interval C++ people

Reply:
Okay. Then I will go with `(const Index* cols_begin, Index num_cols)` where array half-open interval
[cols_begin, cols_begin+num_cols)

Reply:
Keep in mind, on spidermonkey side, as safety property we need:
- To validate `cols_end`: check if it is in wasm memory boundary and `(cols_end - cols_begin)` is positive and multiple of `Index`;
- or, to validate `num_cols`: check if it is positive and `cols_begin + num_cols * sizeof (Index)` is in wasm memory boundary.

Both require about the same about of ops.